### ISSUE TITLE
problems to create ROOT.TF1 from python function in dev3python2 LCG slot  

### ISSUE THREAD

**Issue by VanyaBelyaev** [root-project/root#8152]
- [ ] Checked for duplicates

Cannot create `ROOT.TF1` from python function in  ROOT v6.25/01  (dev3python2  LCG slot) i tworks for dev3 slot, but nor for dev2python2 

### Describe the bug

Failure to create `ROOT.TF1` bject from python function

### Expected behavior

it should work both for python2 and python3 

### To Reproduce

```
from __future__ import print_function
import ROOT, sys,  math  

print( 'Python: ', sys.version_info         )
print( 'ROOT:   ', ROOT.gROOT.GetVersion()  )

def fun ( x,*_ ): return math.sin(x[0])/x[0] 

f1 = ROOT.TF1 ( 'f1' , fun , 0 , 8*math.pi )
f1.Draw ()
```

as output one gets: 

```
Traceback (most recent call last):
  File "./ostap/build/tst1.py", line 10, in <module>
    f1 = ROOT.TF1 ( 'f1' , fun , 0 , 8*math.pi )
TypeError: Template method resolution failed:
  none of the 8 overloaded methods succeeded. Full details:
  TF1::TF1(const TF1& f1) =>
    TypeError: takes at most 1 arguments (4 given)
  TF1::TF1() =>
    TypeError: takes at most 0 arguments (4 given)
  TF1::TF1(const char* name, const char* formula, double xmin, double xmax, const char* option) =>
    TypeError: takes at least 5 arguments (4 given)
  TF1::TF1(const char* name, double xmin, double xmax, int npar, int ndim = 1, TF1::EAddToList addToGlobList = EAddToList::kDefault) =>
    TypeError: could not convert argument 2 (a float is required)
  TF1::TF1(const char* name, const char* formula, double xmin = 0, double xmax = 1, TF1::EAddToList addToGlobList = EAddToList::kDefault, bool vectorize = false) =>
    TypeError: could not convert argument 2 (expected string or Unicode object, function found)
  TF1::TF1(const char* name, ROOT::Math::ParamFunctorTempl<double> f, double xmin = 0, double xmax = 1, int npar = 0, int ndim = 1, TF1::EAddToList addToGlobList = EAddToList::kDefault) =>
    TypeError: could not convert argument 2
  TF1::TF1(const char* name, Double_t(*)(Double_t*,Double_t*) fcn, double xmin = 0, double xmax = 1, int npar = 0, int ndim = 1, TF1::EAddToList addToGlobList = EAddToList::kDefault) =>
    TypeError: could not convert argument 2
  TF1::TF1(const char* name, Double_t(*)(const Double_t*,const Double_t*) fcn, double xmin = 0, double xmax = 1, int npar = 0, int ndim = 1, TF1::EAddToList addToGlobList = EAddToList::kDefault) =>
    TypeError: could not convert argument 2
  Failed to instantiate "TF1(std::string,function,int,double)"
``` 

### Setup

1. dev3ptython2  LCG slot   (ROOT v6.25/01) 
2. lxplus.cern.ch 
3. 
```
( source /cvmfs/sft.cern.ch/lcg/views/LCG_100/${CMTCONFIG}/setup.sh ; python  test.py )
```

### Additional context

It worked for all previous versions of ROOT in all released LCG slots
and it works also for dev3 and decv3python39 LCG slots

**Comment by etejedor** [comment]
Hi @VanyaBelyaev ,

This is strange, we test this every night both in Python2 and Python3:

https://github.com/root-project/root/blob/master/bindings/pyroot/pythonizations/test/tf_pycallables.py

and it works for both.

I can reproduce your error on lxplus with dev3python2, but I can't reproduce on a CentOS7 and an Ubuntu machines (both with master and Python2). Can you reproduce this outside of the LCG stack? It could be an issue with that stack or a particular configuration that we are not testing.

**Comment by VanyaBelyaev** [comment]
Dear @etejedor 
I have no local built of ROOT - it is too heavy for me - I am always rely on LCG releases or nightlies. 
The problem appeared not earlier than a week ago. There was a few days break in tests of my project, 
and restarting tests after this  pause, couple of  days ago, I've observed this behaviour.

**Comment by etejedor** [comment]
I tested with `/cvmfs/sft.cern.ch/lcg/views/dev3python2/Tue/x86_64-centos7-gcc8-opt/`, from May 4th (8 days ago), and I already see the error. But I can't see anything recent before May 4th:

https://github.com/root-project/root/commits/master/bindings/pyroot

that has been committed to PyROOT and could cause this.

I'll check with the librarians about the LCG stack.

**Comment by etejedor** [comment]
I've been able reproduce this on my desktop (CentOS7, gcc 8.3) when building ROOT from source. The parameter that makes it fail is `-DCMAKE_CXX_STANDARD=17` (C++ 14 works fine).

The reason is: when passing a Python callable as a parameter of a function (e.g. the `TF1` constructor in your example) cppyy automatically generates a C++ wrapper for it. As part of the wrapper generation, cppyy jits Python headers, some of which still use (in Python2) the `register` keyword, which causes an error in C++ 17. For example, when running your reproducer with dev3 these errors appear first:

```
In file included from /cvmfs/sft-nightlies.cern.ch/lcg/nightlies/dev3python2/Thu/ROOT/HEAD/x86_64-centos7-gcc8-opt/include/CPyCppyy/API.h:30:
In file included from /cvmfs/sft.cern.ch/lcg/releases/Python/2.7.16-3adfa/x86_64-centos7-gcc8-opt/include/python2.7/Python.h:88:
/cvmfs/sft.cern.ch/lcg/releases/Python/2.7.16-3adfa/x86_64-centos7-gcc8-opt/include/python2.7/unicodeobject.h:534:5: error: ISO C++17 does not allow 'register' storage class specifier [-Wregister]
    register PyObject *obj,     /* Object */
    ^~~~~~~~~
```

this is what ultimately causes the `TF1` constructor error, since the C++ wrapper for the Python callable can't be generated. The errors above start appearing with 6.24 as a result of the LLVM upgrade. But it's legit that they appear, the Python2 headers are just old and not updated anymore.

**Comment by VanyaBelyaev** [comment]
dear @etejedor 
Thank you very much!

**Comment by etejedor** [comment]
No problem!

Fixed changed to https://github.com/root-project/root/pull/8270 plus another PR that will come after the required interface to `TInterpreter` is added.

**Comment by hahnjo** [comment]
@etejedor can we close this issue? (It's already marked as fixed)

**Comment by etejedor** [comment]
@hahnjo it's actually not fixed, I will remove it from the 6.26 project (it still needs a new interface in `TInterpreter`).

**Comment by guitargeek** [comment]
Fixed in ROOT `master` because that version doesn't support Python 2 anymore (Python 2 support was deprecated with ROOT 6.30).
